How many late days are you using for this assignment? 0 late days

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from scipy import stats
#from sklearn.preprocessing import MinMaxScaler

In [2]:
pd.set_option('display.max_columns', 1000)
pd.set_option('display.max_rows', 1000)

# Part 1: Load the dataset

In [3]:
# Load the given datasets
cat_df = pd.read_csv("chronic_kidney_disease_categorical.csv")
num_df = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
display(cat_df)

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
5,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
6,950823,4.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
7,901528,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,433613,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,621332,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,poor,no,no,ckd


In [4]:
display(num_df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.50,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.00,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.80,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.00,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.90,140.0,4.1,NaN,NaN,NaN,NaN
5,992643,35.0,70.0,82.0,36.0,1.10,150.0,3.5,14.5,52.0,9400.0,6.1
6,573400,60.0,80.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,109053,64.0,60.0,106.0,27.0,0.70,150.0,3.3,14.4,42.0,8100.0,4.7
8,352995,33.0,60.0,130.0,41.0,0.90,141.0,4.4,15.5,52.0,4300.0,5.8
9,395450,65.0,70.0,203.0,46.0,1.40,NaN,NaN,11.4,36.0,5000.0,4.1


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: This dataset contains many health-related features, such as specific blood test results and common comorbidities, as well as whether or not the person has chronic kidney disease, in order to predict chronic kidney disease.

The features of this dataset are age, blood pressure (bp, in mm/Hg), specific gravity (sg), albumin (al), sugar (su), red blood cells (rbc), pus cell (pc), pus cell clumps (pcc), bacteria (ba), blood glucose random (bgr, in mgs/dl), blood urea (bu, in mgs/dl), serum creatinine (sc, in mgs/dl), sodium (sod, in mEq/L), potassium (pot, in mEq/L), hemoglobin (hemo, in gms), packed cell volume (pcv), white blood cell count (wbcc, in cells/cmm), red blood cell count (rbcc, in millions/cmm), hypertension (htn), diabetes mellitus (dm), coronary artery disease (cad), appetite (appet), pedal edema (pe), and anemia (ane).

The labels of this dataset are "ckd" and "not ckd" from the class or Target column, indicating whether or not the person has chronic kidney disease.

The dataset has both categorical and numerical features as columns. 3 categorical features are hypertension, appetite, and anemia. 3 numerical features are age, blood pressure, and potassium.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [5]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print("The numerical dataset has " + str(num_df.duplicated(subset=["unique_id"]).sum()) + " duplicate rows.")

# Drop these duplicate rows
num_df = num_df.drop_duplicates(subset=["unique_id"])

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print("The categorical dataset has " + str(cat_df.duplicated(subset=["unique_id"]).sum()) + " duplicate rows.")
cat_df = cat_df.drop_duplicates(subset=["unique_id"])


The numerical dataset has 30 duplicate rows.
The categorical dataset has 30 duplicate rows.


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all of the unique ids are present in both datasets, since the categorical dataset has 10 more unique rows than the numerical dataset (400 vs. 390). The rows that are missing from one of the datasets are not present in the combined table, since merge uses an inner join by default.

In [6]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
both_df = num_df.merge(cat_df, on='unique_id')
#Print the combined dataset
display(both_df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.50,147.0,3.5,13.6,44.0,7300.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.00,144.0,5.0,17.8,43.0,7200.0,5.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.80,135.0,4.1,16.2,50.0,5500.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.00,136.0,4.6,12.0,37.0,8200.0,4.5,0.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.90,140.0,4.1,NaN,NaN,NaN,NaN,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
5,992643,35.0,70.0,82.0,36.0,1.10,150.0,3.5,14.5,52.0,9400.0,6.1,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
6,573400,60.0,80.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,2.0,NaN,NaN,notpresent,notpresent,no,yes,no,good,no,no,ckd
7,109053,64.0,60.0,106.0,27.0,0.70,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,352995,33.0,60.0,130.0,41.0,0.90,141.0,4.4,15.5,52.0,4300.0,5.8,NaN,NaN,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,395450,65.0,70.0,203.0,46.0,1.40,NaN,NaN,11.4,36.0,5000.0,4.1,1.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [7]:
# Calculate the percentage of rows that contain atleast one missing value
proportion = 1 - (len(both_df.dropna()) / len(both_df)) 
# Print %
print("Percent of rows that contain at least one missing value: " + str(100*proportion) + "%")
# Drop these rows from the dataset
both_df = both_df.dropna()
# Print the Dataset
display(both_df)

Percent of rows that contain at least one missing value: 60.769230769230774%


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
13,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
17,630177,58.0,80.0,131.0,18.0,1.1,141.0,3.5,15.8,53.0,6800.0,6.1,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
18,741276,23.0,60.0,95.0,24.0,0.8,145.0,5.0,15.0,52.0,6300.0,4.6,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
19,546010,49.0,80.0,122.0,32.0,1.2,139.0,3.9,17.0,41.0,5600.0,4.9,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
20,460465,47.0,60.0,109.0,25.0,1.1,141.0,4.7,15.8,41.0,8300.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
22,965166,48.0,80.0,75.0,22.0,0.8,137.0,5.0,16.8,51.0,6000.0,6.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [8]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
both_df = both_df.sort_values(by="Target", ignore_index=True)
# Print the dataset
display(both_df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,normal,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,abnormal,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,12100.0,3.7,2.0,0.0,abnormal,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,abnormal,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
5,566924,49.0,100.0,129.0,158.0,11.8,122.0,3.2,8.1,24.0,9600.0,3.5,3.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,no,poor,yes,yes,ckd
6,546225,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,14600.0,3.2,1.0,0.0,normal,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
7,602404,55.0,90.0,273.0,235.0,14.2,132.0,3.4,8.3,22.0,14600.0,2.9,2.0,1.0,abnormal,abnormal,notpresent,notpresent,yes,yes,no,poor,yes,yes,ckd
8,514721,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,no,no,ckd
9,899916,83.0,70.0,102.0,60.0,2.6,115.0,5.7,8.7,26.0,12800.0,3.1,3.0,0.0,normal,normal,notpresent,notpresent,yes,no,no,poor,no,yes,ckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [9]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
both_df = pd.get_dummies(both_df, columns=categorical_columns)
# Print the dataset
display(both_df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,12100.0,3.7,2.0,0.0,True,False,True,False,False,True,True,False,False,True,False,True,True,False,False,True,True,False,False,True,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
5,566924,49.0,100.0,129.0,158.0,11.8,122.0,3.2,8.1,24.0,9600.0,3.5,3.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,True,False,False,True,False,True,False,True,True,False
6,546225,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,14600.0,3.2,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
7,602404,55.0,90.0,273.0,235.0,14.2,132.0,3.4,8.3,22.0,14600.0,2.9,2.0,1.0,True,False,True,False,True,False,True,False,False,True,False,True,True,False,False,True,False,True,False,True,True,False
8,514721,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4,2.0,0.0,True,False,True,False,True,False,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False
9,899916,83.0,70.0,102.0,60.0,2.6,115.0,5.7,8.7,26.0,12800.0,3.1,3.0,0.0,False,True,False,True,True,False,True,False,False,True,True,False,True,False,False,True,True,False,False,True,True,False


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: For a binary categorical variable, having a single column for that variable would be equivalent, because if you only have two values, you can simply assign "True" to one value and "False" to the other - to get the other column, you can just flip the values, so you don't lose any information. If the categorical variable can take more than 2 values, if you try to simplify it to a single binary variable, you lose information, since multiple different values will be grouped together under the "False" label.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [10]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    both_df = both_df[np.abs(stats.zscore(both_df[col])) < 3]
# Print the dataset
both_df = both_df.reset_index(drop=True)
display(both_df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
3,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
4,546225,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,14600.0,3.2,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
5,514721,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4,2.0,0.0,True,False,True,False,True,False,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False
6,621332,59.0,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3,1.0,0.0,True,False,False,True,True,False,True,False,True,False,False,True,True,False,False,True,True,False,True,False,True,False
7,481293,60.0,60.0,288.0,36.0,1.7,130.0,3.0,7.9,25.0,15200.0,3.0,3.0,1.0,False,True,True,False,False,True,True,False,False,True,True,False,True,False,False,True,True,False,False,True,True,False
8,615697,6.0,60.0,94.0,67.0,1.0,135.0,4.9,9.9,30.0,16700.0,4.8,4.0,0.0,True,False,True,False,True,False,False,True,True,False,True,False,True,False,False,True,True,False,True,False,True,False
9,843362,71.0,60.0,118.0,125.0,5.3,136.0,4.9,11.4,35.0,15200.0,4.3,4.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,False,True,False,True,True,False,True,False


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [11]:
# Normalize the all Numerical Attributes in the dataset.
for col in numerical_columns:
    both_df[col] = (both_df[col] - both_df[col].min()) / (both_df[col].max() - both_df[col].min())
# Print the dataset
display(both_df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,0.257143,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,137148,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,0.057143,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,0.057143,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
3,532520,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,0.314286,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
4,546225,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,0.057143,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
5,514721,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,0.169355,0.114286,2.0,0.0,True,False,True,False,True,False,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False
6,621332,0.716216,0.50,1.000000,0.163399,0.111111,0.066667,0.206897,0.267327,0.387097,0.532258,0.371429,1.0,0.0,True,False,False,True,True,False,True,False,True,False,False,True,True,False,False,True,True,False,True,False,True,False
7,481293,0.729730,0.00,0.935622,0.169935,0.160494,0.333333,0.034483,0.019802,0.064516,0.879032,0.000000,3.0,1.0,False,True,True,False,False,True,True,False,False,True,True,False,True,False,False,True,True,False,False,True,True,False
8,615697,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,0.225806,1.000000,0.514286,4.0,0.0,True,False,True,False,True,False,False,True,True,False,True,False,True,False,False,True,True,False,True,False,True,False
9,843362,0.878378,0.00,0.206009,0.751634,0.604938,0.533333,0.689655,0.366337,0.387097,0.879032,0.371429,4.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,False,True,False,True,True,False,True,False


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The column "unique_id" is just an arbitrary id for each person with no medical significance, and thus should not be used for predictions.

In [12]:
#Remove that column
both_df = both_df.drop(columns="unique_id")
# Print the dataset
display(both_df)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,0.257143,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,0.057143,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,0.057143,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
3,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,0.314286,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
4,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,0.057143,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
5,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,0.169355,0.114286,2.0,0.0,True,False,True,False,True,False,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False
6,0.716216,0.50,1.000000,0.163399,0.111111,0.066667,0.206897,0.267327,0.387097,0.532258,0.371429,1.0,0.0,True,False,False,True,True,False,True,False,True,False,False,True,True,False,False,True,True,False,True,False,True,False
7,0.729730,0.00,0.935622,0.169935,0.160494,0.333333,0.034483,0.019802,0.064516,0.879032,0.000000,3.0,1.0,False,True,True,False,False,True,True,False,False,True,True,False,True,False,False,True,True,False,False,True,True,False
8,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,0.225806,1.000000,0.514286,4.0,0.0,True,False,True,False,True,False,False,True,True,False,True,False,True,False,False,True,True,False,True,False,True,False
9,0.878378,0.00,0.206009,0.751634,0.604938,0.533333,0.689655,0.366337,0.387097,0.879032,0.371429,4.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,False,True,False,True,True,False,True,False


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [13]:
# Export the dataframe to a new csv file
both_df.to_csv('cleaned.csv', index=False)  
# Export the dataframe to a new json file
both_df.to_json('cleaned.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [14]:
print(both_df.head(15))

         age    bp       bgr        bu        sc       sod       pot  \
0   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1   0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
2   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
3   0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
4   0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
5   0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.655172   
6   0.716216  0.50  1.000000  0.163399  0.111111  0.066667  0.206897   
7   0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.034483   
8   0.000000  0.00  0.103004  0.372549  0.074074  0.500000  0.689655   
9   0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.689655   
10  0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
11  0.878378  0.25  0.639485  0.470588  0.395062  0.433333  0.517241   
12  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.86

### Paste the markdown table here

Paste here: 
| | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|-----------|-------|-----------|-----------|-----------|-----------|-----------|------------|-----------|-----------|-----------|------|------|---------------|------------|-------------|-----------|----------------|-------------|---------------|------------|--------|---------|-------|--------|--------|---------|-------------|-------------|-------|--------|--------|---------|-------------|----------------|
| 0 | 0.202703 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.206897 | 0.059406 | 0.000000 | 0.653226 | 0.257143 | 4.0 | 0.0 | False | True | True | False | False | True | False | True | True | False | True | False | True | False | True | False | True | False | False | True | True | False |
| 1 | 0.905405 | 1.00 | 0.965665 | 0.522876 | 0.641975 | 0.666667 | 0.000000 | 0.148515 | 0.225806 | 0.217742 | 0.057143 | 3.0 | 2.0 | True | False | True | False | False | True | True | False | False | True | False | True | False | True | False | True | True | False | True | False | True | False |
| 2 | 0.743243 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.000000 | 0.032258 | 0.395161 | 0.057143 | 2.0 | 0.0 | True | False | True | False | True | False | True | False | False | True | False | True | False | True | False | True | False | True | False | True | True | False |
| 3 | 0.729730 | 0.75 | 0.150215 | 0.281046 | 0.234568 | 0.533333 | 0.793103 | 0.336634 | 0.322581 | 0.500000 | 0.314286 | 2.0 | 0.0 | True | False | False | True | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False |
| 4 | 0.540541 | 0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.379310 | 0.207921 | 0.161290 | 0.830645 | 0.057143 | 1.0 | 0.0 | False | True | False | True | True | False | True | False | False | True | False | True | True | False | True | False | True | False | True | False | True | False |
| 5 | 0.675676 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | 0.193548 | 0.169355 | 0.114286 | 2.0 | 0.0 | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False |
| 6 | 0.716216 | 0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667 | 0.206897 | 0.267327 | 0.387097 | 0.532258 | 0.371429 | 1.0 | 0.0 | True | False | False | True | True | False | True | False | True | False | False | True | True | False | False | True | True | True | False | True | True | False |
| 7 | 0.729730 | 0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333 | 0.034483 | 0.019802 | 0.064516 | 0.879032 | 0.000000 | 3.0 | 1.0 | False | True | True | False | False | True | True | False | False | True | True | False | True | False | False | True | True | False | False | True | True | False |
| 8 | 0.000000 | 0.00 | 0.103004 | 0.372549 | 0.074074 | 0.500000 | 0.689655 | 0.217822 | 0.225806 | 1.000000 | 0.514286 | 4.0 | 0.0 | True | False | True | False | True | False | False | True | True | False | True | False | True | False | False | False | True | True | False | True | True | False |
| 9 | 0.878378 | 0.00 | 0.206009 | 0.751634 | 0.604938 | 0.533333 | 0.689655 | 0.366337 | 0.387097 | 0.879032 | 0.371429 | 4.0 | 0.0 | False | True | False | True | True | False | True | False | True | False | True | False | True | False | False | True | True | False | True | False | True | False |
| 10 | 0.851351 | 0.25  | 0.618026 | 0.562092  | 0.728395  | 0.000000   | 0.344828   | 0.168317   | 0.161290  | 0.580645  | 0.085714  | 4.0  | 3.0  | False         | True       | True        | False     | False          | True        | False         | True       | False  | True    | True  | False  | True   | False   | True        | False      | True  | False  | False  | True    | True        | False          |
| 11 | 0.878378 | 0.25  | 0.639485 | 0.470588  | 0.395062  | 0.433333   | 0.517241   | 0.267327   | 0.322581  | 0.104839  | 0.171429  | 3.0  | 0.0  | False         | True       | True        | False     | False          | True        | False         | True       | False  | True    | True  | False  | True   | False   | True        | False      | True  | False  | False  | True    | True        | False          |
| 12 | 0.783784 | 0.00  | 0.725322 | 0.313725  | 0.481481  | 0.566667   | 0.862069   | 0.178218   | 0.193548  | 0.258065  | 0.114286  | 4.0  | 1.0  | True          | False      | True        | False     | True           | False       | False         | True       | False  | True    | True  | False  | True   | False   | False       | True       | True  | False  | False  | True    | True        | False          |
| 13 | 0.662162 | 0.50  | 0.618026 | 0.411765  | 0.432099  | 0.566667   | 0.689655   | 0.316832   | 0.354839  | 0.250000  | 0.200000  | 3.0  | 1.0  | False         | True       | False       | True      | False          | True        | False         | True       | False  | True    | True  | False  | True   | True    | False       | True       | False | False  | True   | False   | True        | False          |
| 14 | 0.770270 | 1.00  | 0.901288 | 0.163399  | 0.345679  | 0.766667   | 0.206897   | 0.524752   | 0.548387  | 0.443548  | 0.342857  | 2.0  | 2.0  | False         | True       | False       | True      | True           | False       | False         | True       | False  | True    | True  | False  | False  | True    | True        | False      | True  | False  | False  | True    | True        | False          |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: You might be able to mitigate hallucinations with prompt engineering, specifically by telling the LLM to double-check its work, or by giving it a few examples of the type of output that you expect (few-shot learning).

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [15]:
# Code to rename all the columns in the dataset
gpt_df = both_df

# Mapping of abbreviations to full names
abbreviation_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia'
}

def rename_columns(col):
    # Split the column into the abbreviation and any suffix
    parts = col.split('_', 1)  # Split at the first underscore only
    prefix = parts[0]           # Get the abbreviation (prefix)
    suffix = parts[1] if len(parts) > 1 else ''  # Get the suffix if it exists

    # Get the full name from the mapping, and add the suffix back if it exists
    full_name = abbreviation_mapping.get(prefix, prefix)  # Default to the original if not found
    return f"{full_name}_{suffix}" if suffix else full_name

# Rename the DataFrame columns
gpt_df.columns = [rename_columns(col) for col in gpt_df.columns]

# Display the renamed DataFrame
display(gpt_df)

,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,wbcc,rbcc,albumin,sugar,red blood cells_abnormal,red blood cells_normal,pus cell_abnormal,pus cell_normal,pus cell clumps_notpresent,pus cell clumps_present,bacteria_notpresent,bacteria_present,hypertension_no,hypertension_yes,diabetes mellitus_no,diabetes mellitus_yes,coronary artery disease_no,coronary artery disease_yes,appetite_good,appetite_poor,pedal edema_no,pedal edema_yes,anemia_no,anemia_yes,Target_ckd,Target_notckd
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,0.257143,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,0.057143,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,0.057143,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
3,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,0.314286,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
4,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,0.057143,1.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False
5,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,0.169355,0.114286,2.0,0.0,True,False,True,False,True,False,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False
6,0.716216,0.50,1.000000,0.163399,0.111111,0.066667,0.206897,0.267327,0.387097,0.532258,0.371429,1.0,0.0,True,False,False,True,True,False,True,False,True,False,False,True,True,False,False,True,True,False,True,False,True,False
7,0.729730,0.00,0.935622,0.169935,0.160494,0.333333,0.034483,0.019802,0.064516,0.879032,0.000000,3.0,1.0,False,True,True,False,False,True,True,False,False,True,True,False,True,False,False,True,True,False,False,True,True,False
8,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,0.225806,1.000000,0.514286,4.0,0.0,True,False,True,False,True,False,False,True,True,False,True,False,True,False,False,True,True,False,True,False,True,False
9,0.878378,0.00,0.206009,0.751634,0.604938,0.533333,0.689655,0.366337,0.387097,0.879032,0.371429,4.0,0.0,False,True,False,True,True,False,True,False,False,True,False,True,True,False,False,True,False,True,True,False,True,False


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [16]:
# Converted SQL to Pandas code
result = gpt_df.groupby('Target_ckd').size().reset_index(name='count')
display(result)

,Target_ckd,count
0,False,111
1,True,22


Answer: The code above groups the dataframe by the Target_ckd column, counts the number of occurrences of each value (in this case, only True and False), then resets the index of the resulting dataframe and gives a name to the count column.